### PR TITLE
Terminate bg worker on drop database

### PR DESCRIPTION
**Pull request by onderkalaci**
I'm working on #1579.

My goal is the following:

* On `DROP DATABASE`, remove the db from the hash and terminate the bg worker before the standard utility kicks in
   * If standard utility succeeds we're done.
   * If fails, we'll re-start the bg worker since we've already removed that.

However, this doesn't work as expected. I couldn't make `TerminateBackgroundWorker () && WaitForBackgroundWorkerShutdown()` actually terminates the worker. This ends up not terminating the session to the database being dropped and getting the same error
```
ERROR:  database "test" is being accessed by other users
DETAIL:  There is 1 other session using the database.
```

Also,  `WaitForBackgroundWorkerShutdown()` really long time. Debugging shows that it actually returns `BGWH_STOPPED`, but on `ps` I can see that worker is not terminated.

Any comments on the approach or the problems I face? What am I missing here?

### CODE REVIEW

File: src/backend/distributed/utils/maintenanced.c [hunk 1]
```diff
@@ -158,7 +184,7 @@ InitializeMaintenanceDaemonBackend(void)
 		memcpy(worker.bgw_extra, &extensionOwner, sizeof(Oid));
 		worker.bgw_notify_pid = MyProcPid;
 
-		if (!RegisterDynamicBackgroundWorker(&worker, &handle))
+		if (!RegisterDynamicBackgroundWorker(&worker, &dbData->workerHandle))
```
Comment (anarazel) [inline src/backend/distributed/utils/maintenanced.c]: That can't work - the pointer is set to process local memory. So other processes can't access it after this.

**Comment by marcocitus** [comment]
I guess the background worker functions expect there to be a process that keeps tracks of the handles (the parallel query group leader).

We could maybe just kill the maintenance daemon directly. We've effectively been doing this in the regression tests by calling `pg_terminate_backend` and it hasn't been a problem. I pushed a commit.

File: src/backend/distributed/utils/maintenanced.c [hunk 2]
```diff
@@ -238,7 +247,6 @@ CitusMaintenanceDaemonMain(Datum main_arg)
 	error_context_stack = &errorCallback;
 
 	/* wire up signals */
-	pqsignal(SIGTERM, die);
```
Comment (anarazel) [inline src/backend/distributed/utils/maintenanced.c]: Why did you move this? Until the BackgroundWorkerUnblockSignals signals are blocked...

**Comment by onderkalaci** [comment]
> We could maybe just kill the maintenance daemon directly.

This was my next thing to try.

One note on this approach is that if `DROP DATABASE ` fails due to other connections, we'd already have killed the daemon.  And until a new backend starts, we wouldn't be able to start the daemon on the existing backends (since the existing backends would always see `MetadataCache.extensionLoaded` as true, they'll never start the bg worker again).

I think there is not an easy way of fixing that by invalidating cache etc. We could may be not stop bg worker in the first place,  manually call `dropdb()` on `StopMaintenanceDaemon ` and then skip `standardProcessUtility`. 

Or, we can choose to live with that, which doesn't seem too bad to me.

**Comment by anarazel** [comment]
On 2017-08-18 07:52:53 +0000, Önder Kalacı wrote:
> Or, we can choose to live with that, which doesn't seem too bad to me.

+1. The database was about to be dropped after all.

File: src/backend/distributed/utils/maintenanced.c [hunk 3]
```diff
@@ -275,6 +283,9 @@ CitusMaintenanceDaemonMain(Datum main_arg)
 			/*
 			 * We don't want to run the deadlock checks if there exists
 			 * any version mistmatch.
+			 *
+			 * We also want to make sure that the extension is avaliable by checking
```
Comment (onderkalaci) [inline src/backend/distributed/utils/maintenanced.c]: a minor note: I think this is obsolete and will conflict with #1584.

File: src/test/regress/expected/multi_extension.out [hunk 4]
```diff
@@ -287,15 +287,6 @@ FROM test.maintenance_worker();
...
 \c regression
-SELECT datname,
-    pg_terminate_backend(pid)
-FROM test.maintenance_worker('another');
- datname | pg_terminate_backend 
----------+----------------------
- another | t
-(1 row)
-
 DROP DATABASE another;
```
Comment (onderkalaci) [inline src/test/regress/expected/multi_extension.out]: Shall we run 

```
 SELECT datname,
     datname = 'another',		    
     usename = (SELECT extowner::regrole::text FROM pg_extension WHERE extname = 'citus'
 FROM test.maintenance_worker();	

```
to make sure that the daemon is dead?

File: src/backend/distributed/utils/maintenanced.c [hunk 5]
```diff
@@ -86,6 +89,7 @@ static size_t MaintenanceDaemonShmemSize(void);
 static void MaintenanceDaemonShmemInit(void);
 static void MaintenanceDaemonErrorContext(void *arg);
 
+
```
Comment (onderkalaci) [inline src/backend/distributed/utils/maintenanced.c]: extra new line

Comment (onderkalaci) [inline src/backend/distributed/utils/maintenanced.c]: Oh not, just fixing it

**Comment by marcocitus** [comment]
We might want to unblock signals before releasing the lwlock. That way we're guaranteed that the signal handler is set up when we do a kill.

File: src/backend/distributed/utils/maintenanced.c (hunk 2 above)
Comment (anarazel) [inline src/backend/distributed/utils/maintenanced.c]: Ping, why are we moving the pqsignal around?

**Comment by anarazel** [comment]
> We might want to unblock signals before releasing the lwlock. That way we're guaranteed that the signal handler is set up when we do a kill.

Hm? A blocked signal arriving while blocked will be delivered when the signals are unblocked, including using the right handler.

**Comment by marcocitus** [comment]
> Hm? A blocked signal arriving while blocked will be delivered when the signals are unblocked, including using the right handler.

We'd like to the process to die as soon as the signal arrives though.

**Comment by onderkalaci** [comment]
@marcocitus should be ready for you to look

File: src/backend/distributed/utils/maintenanced.c [hunk 6]
```diff
@@ -243,7 +251,6 @@ CitusMaintenanceDaemonMain(Datum main_arg)
 	error_context_stack = &errorCallback;
 
 	/* wire up signals */
-	pqsignal(SIGTERM, die);
 	pqsignal(SIGHUP, MaintenanceDaemonSigHupHandler);
 	BackgroundWorkerUnblockSignals();
```
Comment (marcocitus) [inline src/backend/distributed/utils/maintenanced.c]: This two lines should move up to the other pqsignal call.

File: src/backend/distributed/utils/maintenanced.c [hunk 7]
```diff
@@ -286,6 +293,14 @@ CitusMaintenanceDaemonMain(Datum main_arg)
 		{
 			StartTransactionCommand();
 
+			/*
+			 * We skip the deadlock detection if citus extension
+			 * is not accessible.
+			 *
+			 * Similarly, we skip to run the deadlock checks if
+			 * there exists any version mistmatch or the extension
```
Comment (marcocitus) [inline src/backend/distributed/utils/maintenanced.c]: mismatch

File: src/test/regress/expected/multi_extension.out [hunk 8]
```diff
@@ -287,15 +288,47 @@ FROM test.maintenance_worker();
...
+    RETURNS pg_stat_activity
+    LANGUAGE plpgsql
+AS $$
+DECLARE
+   activity record;
+   tryCount int := 0;
+BEGIN
+    LOOP
+        SELECT * INTO activity FROM pg_stat_activity
+    WHERE application_name = 'Citus Maintenance Daemon' AND datname = p_dbname;
```
Comment (marcocitus) [inline src/test/regress/expected/multi_extension.out]: alignment is off

File: src/test/regress/expected/multi_extension.out [hunk 9]
```diff
@@ -287,15 +288,47 @@ FROM test.maintenance_worker();
...
-SELECT datname,
-    pg_terminate_backend(pid)
-FROM test.maintenance_worker('another');
- datname | pg_terminate_backend 
----------+----------------------
- another | t
+CREATE SCHEMA test_deamon;
+-- we have to create the same function on the regression database as well
+-- note that this function is slightly different then the above, where
+-- the function tries for 5 times then exits if no deamon is found
```
Comment (marcocitus) [inline src/test/regress/expected/multi_extension.out]: why not just loop until the daemon is gone?

Comment (onderkalaci) [inline src/test/regress/expected/multi_extension.out]: Makes sense

File: src/backend/distributed/utils/maintenanced.c (hunk 7 above)
Comment (onderkalaci) [inline src/backend/distributed/utils/maintenanced.c]: I'm very consistent about making typos 😄

File: src/test/regress/sql/multi_extension.sql [hunk 10]
```diff
@@ -262,15 +262,52 @@ BEGIN
...
+        IF activity.pid IS NULL THEN
+            RETURN true;
+        ELSE
+            PERFORM pg_sleep(0.1);
+            PERFORM pg_stat_clear_snapshot();
+        END IF;
+        tryCount = tryCount + 1;
+        IF tryCount > 5 THEN
+            RETURN false;
+        END IF;
```
Comment (marcocitus) [inline src/test/regress/sql/multi_extension.sql]: Do we still need the tries?

Comment (onderkalaci) [inline src/test/regress/sql/multi_extension.sql]: Not for this specific test, but might be relevant if the `SIGTERM` is sent to the worker when it is not accepting signals?

Comment (marcocitus) [inline src/test/regress/sql/multi_extension.sql]: Do we have a bug that allows that? 

It seems like we can only have two cases:
1) `StopMaintenanceDaemon` runs before the worker acquires the lwlock. The worker sees its not in the hash and calls `proc_exit(0)`.
2) `StopMaintenanceDaemon` runs after the worker acquires the lwlock. The signals and `workerPid` set up and when StopMaintenanceDaemon calls `kill`, the worker calls `die`.

Comment (onderkalaci) [inline src/test/regress/sql/multi_extension.sql]: > Do we have a bug that allows that?

That's right, I got it now. We should probably not have the loop to test the above as well.
